// @botpress/builtins/dist/renderers/builtin_single-choice.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _url = require('url');

var _url2 = _interopRequireDefault(_url);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = data => [{
  on: 'facebook',
  text: data.text,
  quick_replies: data.choices.map(c => `<${c.value}> ${c.title}`),
  typing: data.typing
}, {
  on: 'webchat',
  text: data.text,
  quick_replies: data.choices.map(c => `<${c.value}> ${c.title}`),
  typing: data.typing
}, {
  on: 'microsoft',
  type: 'message',
  text: data.text,
  inputHint: 'expectingInput',
  suggestedActions: {
    actions: data.choices.map(c => ({
      type: 'imBack',
      title: c.title,
      value: c.value
    }))
  }
}, {
  on: 'slack',
  attachments: [{
    text: data.text,
    attachment_type: 'default',
    actions: data.choices.map(c => ({
      name: 'press',
      text: c.title,
      type: 'button',
      value: c.value
    }))
  }]
}]; // TODO
// Add buttons support to Telegram
